Bookmarks migration targeting bookmarks feature branch

adam-james-v opened the pull request:
# Bookmarks BE

This PR contains:

- migrations to add 3 Tables:
  - card_bookmark
  - dashboard_bookmark
  - collection_bookmark
- bookmark models
  - CardBookmark
  - DashboardBookmark
  - CollectionBookmark
  - `metabase.models.bookmark` with fn `bookmarks-for-user` to fetch bookmarks from each table at once
- bookmark API
  - [x] GET /api/bookmark which covers issue #20527
  - [x] POST, DELETE /api/bookmark/:model:/:id for issue #20388 (also note the next 3 endpoints also allow DELETE)
  - [x] POST, DELETE /api/card/:id/bookmark
  - [ ] POST, DELETE /api/collection/root|:id/bookmark
  - [x] POST, DELETE /api/dashboard/:id/bookmark
- hydrating `is_bookmarked` key on cards, dashboards, collections
  - [x] GET /api/card/:id
  - [ ] GET /api/dashboard/:id
  - [ ] GET /api/collection/:id

dpsutton committed:
Bookmark tables

- card_bookmarks
- dashboard_bookmarks
- collection_bookmarks

All of shape (id, user_id, {card,dashboard,collection}_id, created_at
indices on user_id, item_id
unique constraint on (user_id, item_id)

dpsutton committed:
Bookmark tables

- card_bookmarks
- dashboard_bookmarks
- collection_bookmarks

All of shape (id, user_id, {card,dashboard,collection}_id, created_at
indices on user_id, item_id
unique constraint on (user_id, item_id)

dpsutton committed:
Bookmark models and api

dpsutton committed:
Bookmark models and api

dpsutton committed:
Namespace doc and fixup POST endpoint

dpsutton committed:
Namespace doc and fixup POST endpoint

adam-james-v committed:
Fixed a few typos

adam-james-v committed:
Fixed a few typos

adam-james-v committed:
Merge branch 'bookmarks-migration' of https://github.com/metabase/metabase into bookmarks-migration

adam-james-v committed:
Made a WIP implementation for grabbing all bookmarks for the user

Using a honeysql query, it's possible to union the results from the tables:

- card_bookmark
- dashboard_bookmark
- collection_bookmark

Where the user_id is the current user.

Then, we can left join this table (called 'bookmark' using :with) with whatever tables make sense. For example, we may
care about report_card and card table data joined with the card_id, same for dashboards, and collections.

The shape of the returned data is not super clean yet, but the query works as a start.

adam-james-v committed:
Made a WIP implementation for grabbing all bookmarks for the user

Using a honeysql query, it's possible to union the results from the tables:

- card_bookmark
- dashboard_bookmark
- collection_bookmark

Where the user_id is the current user.

Then, we can left join this table (called 'bookmark' using :with) with whatever tables make sense. For example, we may
care about report_card and card table data joined with the card_id, same for dashboards, and collections.

The shape of the returned data is not super clean yet, but the query works as a start.

adam-james-v committed:
Hydrate cards with :is_bookmarked on the GET /api/card/:id endpoint.

Add :is_bookmarked key to the body returned by the endpoint.

adam-james-v committed:
Hydrate cards with :is_bookmarked on the GET /api/card/:id endpoint.

Add :is_bookmarked key to the body returned by the endpoint.

adam-james-v committed:
Bookmarks WIP, changing 'Favorites' references to bookmarks

Starting with the card endpoint, I'm changing over all references of 'favorites' to 'bookmark'. This includes changing
some of the required namespaces and adjusting the logic of a few functions (eg. hydrate :is_bookmarked)

Can change the approach of these as things move forward, but the basics are starting to make sense and solidify here.

adam-james-v committed:
Bookmarks WIP, changing 'Favorites' references to bookmarks

Starting with the card endpoint, I'm changing over all references of 'favorites' to 'bookmark'. This includes changing
some of the required namespaces and adjusting the logic of a few functions (eg. hydrate :is_bookmarked)

Can change the approach of these as things move forward, but the basics are starting to make sense and solidify here.

adam-james-v committed:
rebased from master, fixing conflicts

adam-james-v committed:
Improve the shape of the bookmarks list at GET /api/bookmark

Using the left join in the query, it's possible to get all of the needed details and alias the columns to remove
ambiguity. Then, just filter the map's values, removing keys whose value is nil.

adam-james-v committed:
Getting the Bookmark Endpoint to produce desired shape of data

The bookmark endpoint now normalizes the results and annotates with a :type key for each bookmark.
The bookmarks are sorted by created_at in the BE, but the created_at key is not returned, at least for now. User
preference based ordering is TBD. It might be a FE thing, but not sure.

As well, dashboards and cards are functional with POST, DELETE endpoints and hydrating :is_bookmarked.

Really need to finish collections and then build good tests plus do some general cleanup.

dpsutton commented on src/metabase/models/bookmark.clj:
```diff
@@ -0,0 +1,81 @@
+(ns metabase.models.bookmark
+  (:require [clojure.string :as str]
+            [metabase.db.connection :as mdb]
+            [metabase.models :refer [Card Dashboard Collection]]
```
can't refer to `metabase.models` because that ns refers to this one.

dpsutton commented:
> /work/src/metabase/api/bookmark.clj:18:1: warning: Missing docstring.

adam-james-v committed:
Fixing Errors to get things working

This doesn't necessarily fix the logic of some tests, but it's a first pass getting things at least passing linters
and not failing compilation.

adam-james-v committed:
Fixed yaml issue

adam-james-v committed:
Oops. Fixing yaml again. Made silly mistake.

I remembered the ./bin/lint-migrations-file.sh linting tool, so this should hopefully be a rare mistake :/

adam-james-v committed:
Return id on bookmarks list, required for FE

Started changing search test. Still failing.

adam-james-v committed:
Small change to cmd.copy, removing DashboardFavorite reference

adam-james-v committed:
Fixing tests enough to (hopefully) pass.

adam-james-v committed:
Attempting to get bookmarks test to pass in CI

Not sure what's wrong. This is just seeing if something changes
